ipcc/src/handle.rs

```diff
@@ -0,0 +1,108 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+use std::{
+    ffi::{c_int, CStr, CString},
+    ptr,
+};
+
+use crate::IpccError;
+use crate::{ffi::*, IpccErrorInner};
+
+pub struct IpccHandle(*mut libipcc_handle_t);
+
+impl Drop for IpccHandle {
+    fn drop(&mut self) {
+        unsafe {
+            libipcc_fini(self.0);
+        }
+    }
+}
+fn ipcc_fatal_error<C: Into<String>>(
+    context: C,
+    lerr: libipcc_err_t,
+    syserr: c_int,
+    errmsg: CString,
+) -> IpccError {
+    let context = context.into();
+    let syserr = if syserr == 0 {
+        "no system errno".to_string()
+    } else {
+        let syserr = unsafe { libc::strerror(syserr) };
+        unsafe { CStr::from_ptr(syserr) }.to_string_lossy().into_owned()
+    };
+    let inner = IpccErrorInner {
+        context,
+        errmsg: errmsg.to_string_lossy().into_owned(),
+        syserr,
+    };
+    match lerr {
+        LIBIPCC_ERR_OK => panic!("called fatal on LIBIPCC_ERR_OK"),
+        LIBIPCC_ERR_NO_MEM => IpccError::NoMem(inner),
+        LIBIPCC_ERR_INVALID_PARAM => IpccError::InvalidParam(inner),
+        LIBIPCC_ERR_INTERNAL => IpccError::Internal(inner),
+        LIBIPCC_ERR_KEY_UNKNOWN => IpccError::Internal(inner),
+        LIBIPCC_ERR_KEY_BUFTOOSMALL => IpccError::KeyBufTooSmall(inner),
+        LIBIPCC_ERR_KEY_READONLY => IpccError::KeyReadonly(inner),
+        LIBIPCC_ERR_KEY_VALTOOLONG => IpccError::KeyValTooLong(inner),
+        LIBIPCC_ERR_KEY_ZERR => IpccError::KeyZerr(inner),
+        _ => IpccError::UnknownErr(inner),
+    }
+}
+
+impl IpccHandle {
+    pub fn new() -> Result<Self, IpccError> {
+        let mut ipcc_handle: *mut libipcc_handle_t = ptr::null_mut();
+        let errmsg = CString::new(vec![1; 1024]).unwrap();
```

Comment:
It isn't back yet but libipcc.h will have `#define LIBIPCC_ERR_LEN 1024` - I'd suggest defining this in ffi.rs and using it here.
See https://code.oxide.computer/c/illumos-gate/+/296/14/usr/src/lib/libipcc/common/libipcc.h#46

Reply:
Fixed in 2e25d56